core/modules/xrdsvc/SsiRequest.cc

```diff
@@ -89,56 +91,121 @@ void SsiRequest::execute(XrdSsiRequest& req) {
+                t.stop();
+                LOGS(_log, LOG_LVL_DEBUG, "Enqueued TaskMsg for " << ru << " in " << t.getElapsed() << " seconds");
+            }
+            break;
+
+        case ResourceUnit::WORKER:
+            {
+                if (!(*_validator)(ru)) {
+                    std::ostringstream os;
+                    os << "WARNING: unowned worker detected:" << ru.path();
+                    LOGS(_log, LOG_LVL_WARN, os.str());
+                    errorFunc(os.str());
+                    return;
+                }
+
+                // reqData has the entire request, so we can unpack it without waiting for
+                // more data.
+                LOGS(_log, LOG_LVL_DEBUG, "Decoding WorkerCmdMsg of size " << reqSize);
+                auto workerCmdMsg = std::make_shared<proto::WorkerCmdMsg>();
+                bool ok = workerCmdMsg->ParseFromArray(reqData, reqSize) && workerCmdMsg->IsInitialized();
+                if (!ok) {
+                    std::ostringstream os;
+                    os << "Failed to decode WorkerCmdMsg on worker=" << ru.hashName();
+                    LOGS(_log, LOG_LVL_ERROR, os.str());
+                    errorFunc(os.str());
+                    return;
+                }
+                LOGS(_log, LOG_LVL_INFO, "WorkerCmdMsg: cmd=" << proto::WorkerCmdMsg_Cmd_Name(workerCmdMsg->cmd()));
+
+            
+                // Now that the request is decoded (successfully or not), release the
+                // xrootd request buffer. To avoid data races, this must happen before
+                // the command is handed off to another thread for processing, as there is a
+                // reference to this SsiRequest inside the reply channel for the task,
+                // and after the call to BindRequest.
+                wbase::WorkerCommand::Ptr command;
+                switch (workerCmdMsg->cmd()) {
+
+                    case proto::WorkerCmdMsg::RELOAD_CHUNK_LIST:
+                        command = std::make_shared<wcontrol::ReloadChunkListCommand>(replyChannel);
+                        break;
+
+                    default:
+                        {
+                            std::ostringstream os;
+                            os  << "Unsupported command " << proto::WorkerCmdMsg_Cmd_Name(workerCmdMsg->cmd())
+                                << " found in WorkerCmdMsg on worker=" << ru.hashName();
+                            LOGS(_log, LOG_LVL_ERROR, os.str());
+                            errorFunc(os.str());
+                            return;
+                        }
+                }
+                ReleaseRequestBuffer();
+                t.start();
+                _processor->processCommand(command);    // Queues the command to be run later.
+                t.stop();
+                LOGS(_log, LOG_LVL_DEBUG, "Enqueued WorkerCommand for " << ru << " in " << t.getElapsed() << " seconds");
+            }
+            break;
```

Comment:
It's probably worth considering putting the non-default case code in their own functions. There's a fair bit of similar code in both of these that can probably be put in separate functions as well.